### DISCUSSION TITLE
Registering callbacks to multiple socketio clients in single thread

### DISCUSSION THREAD

**Discussion by maxamel** [miguelgrinberg/python-engineio#381] [General]
Hi, I wanted to ask how registering callbacks to multiple socketio clients works. I have a code snippet where I'm initializing socketio clients in a loop and then emitting a call from each client to the server, whilst registering a callback. In the callback the sid returned is the same sid (presumably for the last client only). When I try the same with a client per thread it works, but for my purposes a client per thread is not feasible so I wanted to understand if this is some kind of known limitation. I saw this line in the docs: _Callback functions can only be used when addressing an individual client._ but it does not seem to apply to my case since I have a unique client per call. I am using latest version of the library.

Code snippet:

```
import time
import socketio


UTILS = 'UTILS'
SIO = 'SIO'
CALLBACK = 'CALLBACK'


class UtilityHelper:

    def __init__(self):
        self.player_sid = None


def test(endpoint: str, sid_to_data: dict):

        try:
            for i in range(4):

                sio = socketio.SimpleClient(logger=False, engineio_logger=False)
                sio.connect(url=endpoint, namespace='/connect', transports=['websocket'])
                assert sio.sid is not None
                print(f'Got new sid {sio.sid}')
                sid_to_data[sio.sid] = dict()
                sid_to_data[sio.sid][UTILS] = UtilityHelper()
                sid_to_data[sio.sid][SIO] = sio

                @sio.client.on('game', namespace='/connect')
                def game(data):
                    print(f'Game event received for {sio.sid}')
                    target_sid = sio.sid
                    util = sid_to_data[target_sid][UTILS]
                    if isinstance(data, str):
                        util.player_sid = data

                sid_to_data[sio.sid][CALLBACK] = game
                #sid_to_data[sio.sid][SIO].client.on('game', game)

            for sid_data in sid_to_data.values():
                util = sid_data[UTILS]
                sio = sid_data[SIO]
                util.player_sid = sio.client.emit(event='/api/play', data={'data': {'preferences': {'time_control': '3+0', }}},
                                                  namespace='/connect', callback=sid_data[CALLBACK])
                print(f'Emmitted play for this player sid {sio.sid}')

            time.sleep(10)

        except Exception as e:
            print(f'Error recorded {e}')


if __name__ == '__main__':
    test('http://localhost:1443/', dict())

```

Output for the above:

```
Got new sid X0pTQKokwZorzqU9AAAK
Got new sid u6rK8XL8DaGbrGGrAAAM
Got new sid CgnSlCiP_6wudk0CAAAJ
Got new sid FIxokmVlhhlj04mLAAAO
Emmitted play for this player sid X0pTQKokwZorzqU9AAAK
Emmitted play for this player sid u6rK8XL8DaGbrGGrAAAM
Emmitted play for this player sid CgnSlCiP_6wudk0CAAAJ
Emmitted play for this player sid FIxokmVlhhlj04mLAAAO
Game event received for FIxokmVlhhlj04mLAAAO
Game event received for FIxokmVlhhlj04mLAAAO
Game event received for FIxokmVlhhlj04mLAAAO
Game event received for FIxokmVlhhlj04mLAAAO
```

**Comment by miguelgrinberg** [comment]
You are mixing up two different types of clients. Do you want to use the simple client or the regular client? The simple client does not use event handlers nor callbacks. This hybrid approach that combines the two clients you are trying to use here is not a supported usage.

So first of all, switch to the regular client (`socketio.Client`). Then I would suggest you define your handlers before connecting, to avoid race conditions. Once your handlers are in place, then connect. Finally, you are providing an event handler as a callback in your emit, which is  incorrect. Use a regular function as a callback, configured with the arguments that your callback needs.

**Comment by maxamel** [comment]
Thanks @miguelgrinberg , I implemented the suggested changes. I concluded that I don't need the callback since from the server I am emitting an event back to the client. So I am using an event handler per client instance. However, I am still seeing the event triggered just for one sid (presumably the last one handled):

Updated code:
```
import time
import socketio


UTILS = 'UTILS'
SIO = 'SIO'
CALLBACK = 'CALLBACK'


class UtilityHelper:

    def __init__(self):
        self.player_sid = None


def test(endpoint: str, sid_to_data: dict):

        try:
            for i in range(4):

                sio = socketio.Client(logger=False, engineio_logger=False)

                def game(data):
                    print(f'Game event received for {sio.sid} and {data}')
                    target_sid = sio.sid
                    util = sid_to_data[target_sid][UTILS]
                    if isinstance(data, str):
                        util.player_sid = data

                sio.connect(url=endpoint, namespaces='/connect', transports=['websocket'])
                assert sio.sid is not None
                print(f'Got new sid {sio.sid}')
                sid_to_data[sio.sid] = dict()
                sid_to_data[sio.sid][UTILS] = UtilityHelper()
                sid_to_data[sio.sid][SIO] = sio
                sid_to_data[sio.sid][SIO].on('game', game, namespace='/connect')

            for sid_data in sid_to_data.values():
                sio = sid_data[SIO]
                sio.emit(event='/api/play', data={'data': {'preferences': {'time_control': '3+0', }}},
                         namespace='/connect')
                print(f'Emitted play for this player sid {sio.sid}')

            time.sleep(10)

        except Exception as e:
            print(f'Error recorded {e}')


if __name__ == '__main__':
    test('http://localhost:1443/', dict())
```

The output:
```
Got new sid HKFwT4430KvcJfAzAAAM
Got new sid hb0EEiw57_WMnLEIAAAU
Got new sid YRz_t7HYTaQzIOjXAAAO
Got new sid YiYULM3H25nhqhB2AAAQ
Emitted play for this player sid HKFwT4430KvcJfAzAAAM
Emitted play for this player sid hb0EEiw57_WMnLEIAAAU
Emitted play for this player sid YRz_t7HYTaQzIOjXAAAO
Emitted play for this player sid YiYULM3H25nhqhB2AAAQ
Game event received for YiYULM3H25nhqhB2AAAQ and {'sid': 'b96b988568934ad4b3912aefc3b5d660'}
Game event received for YiYULM3H25nhqhB2AAAQ and {'sid': 'b087b4404a5044ca87fecc439ce47b90'}
Game event received for YiYULM3H25nhqhB2AAAQ and {'sid': 'b4acf3e4b48e41f6a01d0385eac32d86'}
Game event received for YiYULM3H25nhqhB2AAAQ and {'sid': 'd5d20e5ab0e54466873a8fcf061dc546'}
```
In a socketio-client per thread setup this works fine, so I am assuming there's something wrong with the client-side, and not something on the server-side.

**Reply by miguelgrinberg** [reply]
Your print statement is wrong. You are printing the value of a variable that you reused four times, so obviously the last value assigned to it will print. If the data that your server sends you does not allow you to identify the client that received the event, then you can use `functools.partial` to add the client as an extra argument to your function.

Also note that you are still setting the event handler after connecting, which is not ideal.

**Reply by maxamel** [reply]
Thanks, I was sure there was an easy was to get hold of the sid when dealing with multiple socketio clients in a single thread. But it's OK, I managed to get it working using the _functools.partial_ utility as suggested. Thank you for all your efforts!
